### PR TITLE
ADC raw values calibration

### PR DESCRIPTION
**Pull request by katyo**
Based on #550 
Related issue #326

This PR implements ADC calibration based on ESP IDF. The following schemes supported:
- Basic calibration which sets initial (zero) bias
- Line fitting which corrects gain (also includes basic)
- Curve fitting which corrects non-linearity (also inclues basic and line fitting)

First two schemes uses factory values stored in efuse fields when it available.
As a fallback a runtime calibration is made by connecting ADC input to ground or reference voltage internally.

Third scheme uses pre-defined polynomial coefficients derived from IDF.
Also applied useful optimizations for runtime calculations and to coefficients representation to reduce operations and avoid division.

NOTE: Currently calibration supported only for riscv chips because I haven't xtensa-based boards for testing, but it will be relatively easy add support for such chips later.

### Must

- [x] The code compiles without `errors` or `warnings`.
- [x] All examples work.
- [x] `cargo fmt` was run.
- [x] Your changes were added to the `CHANGELOG.md` in the proper section.
- [x] You updated existing examples or added examples (if applicable).
- [ ] Added examples are checked in CI

### Nice to have

- [x] You add a description of your work to this PR.
- [x] You added proper docs for your newly added features and code.

### CODE REVIEW

File: esp32c3-hal/examples/adc_cal.rs
Comment (bugadani) [inline esp32c3-hal/examples/adc_cal.rs:58]: Why do you have to set 3 different calibration types? What happens if I just define `type AdcCal = (AdcCalCurve<ADC1>);`?

**Comment by JurajSadel** [comment]
@katyo Thank you very much for working on this, is it something I can help you at this stage?

**Comment by katyo** [comment]
@JurajSadel,
Currently I search a way to avoid breaking changes for an API.
We need find a proper way to select calibration/fitting algorithm when enabling ADC pin.

File: esp-hal-common/src/soc/esp32c3/efuse.rs
Comment (bugadani) [inline esp-hal-common/src/soc/esp32c3/efuse.rs:70]: I personally don't like reading Efuses this low level. I think the abstraction [here](https://github.com/dimpolo/esp-hal/commit/f79432469f60df636e00ad80077e27f7b21f9c14#diff-c262537b65a1612b4c12352b3dcfdf9dadbbe6552dedf6250480aea2082071edR213) is easier to read, write, and compare to esp-idf data.

I'm not sure what you think about it, and why you chose reading the registers directly. They are probably faster and they may generate less code, but I don't think this is very critical.

I can refactor the efuses later if you don't have strong feelings against it.

**Comment by bugadani** [comment]
I know my opinion here isn't very important, but I'm pretty happy with where this is heading. From a pure design standpoint, I would be happier if the algorithms weren't tied to the efuses, but having calibration coupled to the ADCs here isn't bad either.

I am slightly worried about the hardcoded constants in curve fitting - how confident can we be that a single set of constants work well across different production runs?

File: esp-hal-common/src/soc/esp32c3/efuse.rs
Comment (katyo) [reply esp-hal-common/src/soc/esp32c3/efuse.rs:70]: Currently we lacks unified way of getting fields from efuse.
The existing methods such as `Efuse::get_mac_address` `Efuse::get_flash_encryption` also works via PAC.

**Comment by katyo** [comment]
@bugadani 
I share your concerns about hardcoded curve coefficients which derived from ESP IDF source. I'm far away from chip manufacturing but curve-based error correction gives me best precision at the moment particularly near to top edge of a range. Seems coefficients represents non-linearity of attenuator and convertor itself.

**Comment by katyo** [comment]
Now this PR uses efuse fields access from #567.

**Review by JurajSadel — COMMENTED** [review]
Thank you very much for working on this!

I tried it, the C3 is working fine!

C6 example (`adc_cal`) is panicking for me with
```rust
!! A panic occured in 'C:\Users\JurajSadel\rust_test\esp-hal\esp-hal-common\src\analog\adc\cal_line.rs', at line 34, column 46

PanicInfo {
    payload: Any { .. },
    message: Some(
        No reference point V1 found in efuse,
    ),
    location: Location {
        file: "C:\\Users\\JurajSadel\\rust_test\\esp-hal\\esp-hal-common\\src\\analog\\adc\\cal_line.rs",
        line: 34,
        col: 46,
    },
    can_unwind: true,
}

Backtrace:

No backtrace available - make sure to force frame-pointers. (see https://crates.io/crates/esp-backtrace)
```


On C2, I _think_ the voltage values are not correct (compared with my toy voltmeter and would be nice if someone could give it a try) and I couldn't compare it with the ESP-IDF because of revision mismatch.

[C2](https://github.com/esp-rs/esp-hal/actions/runs/5180310374/jobs/9334329232?pr=555) and [C6](https://github.com/esp-rs/esp-hal/actions/runs/5180310374/jobs/9334329997?pr=555) have a bunch of unused constants warnings, would be nice to clean them as well

**Comment by katyo** [comment]
@JurajSadel Could you test it again?
I fixed calibration data version detection.
Also I added as a fallback the reference voltage calibration for linear scheme. But tha last one is not so good because actual reference voltage may varies in range 1.0..=1.2V. This value currently is not present in efuse.

**Comment by jessebraham** [comment]
I've confirmed that the examples build and run successfully on the C2, C3, and C6. The values for the C2 and C3 looked correct, however for me the C3 was reporting closer to 2.8V. Other than that seems good, though.

**Comment by bjoernQ** [comment]
I just tried it and it works fine for me on C2 and C3. However C6 crashes in development mode with this
```
!! A panic occured in 'C:\projects\review\katyo\esp-hal\esp-hal-common\src\analog\adc\cal_line.rs', at line 73, column 15

PanicInfo {
    payload: Any { .. },
    message: Some(
        attempt to divide by zero,
    ),
    location: Location {
        file: "C:\\projects\\review\\katyo\\esp-hal\\esp-hal-common\\src\\analog\\adc\\cal_line.rs",
        line: 73,
        col: 15,
    },
    can_unwind: true,
}
```

And in release-mode it certainly shows weird values because of this

**Comment by katyo** [comment]
@bjoernQ Looks strange.

Seems `code` is zero here:
```rust
let gain = ((mv as u32 * GAIN_SCALE * 2 / code as u32 + 1) * 4096 / atten.ref_mv() as u32 + 1) / 2;
```

May be calibration data is missing (or version check failed)
or (a fallback way) runtime calibration failed.

**Comment by bjoernQ** [comment]
Maybe adding some debug or trace logs might help identifying what is going on here. We already have the `log` dependency in `esp-hal-common`

**Comment by elpiel** [comment]
I'm getting this `panic` from embassy when I've changed my patch in Cargo.toml to test this out.

The only change I did was changing from the latest commit of this PR (https://github.com/esp-rs/esp-hal/pull/550#issuecomment-1587102788)  to the latest commit in this branch

```
!! A panic occured in '/home/elpiel/.cargo/registry/src/github.com-1ecc6299db9ec823/embassy-executor-0.2.0/src/raw/mod.rs', at line 355, column 30
 
PanicInfo {
    payload: Any { .. },
    message: Some(
        unwrap of `driver::allocate_alarm()` failed: NoneError,
    ),
    location: Location {
        file: "/home/elpiel/.cargo/registry/src/github.com-1ecc6299db9ec823/embassy-executor-0.2.0/src/raw/mod.rs",
        line: 355,
        col: 30,
    },
    can_unwind: true,
}

```

**Comment by katyo** [comment]
@elpiel Could you try `adc_cal` example?

**Comment by elpiel** [comment]
Example runs successfully for esp32-c3 adc_cal but I still get a panic in embassy.

As for the values, they still look unstable:

```
PIN3 ADC reading = 1102 (1050 mV)
PIN3 ADC reading = 260 (247 mV)
PIN3 ADC reading = 482 (459 mV)
PIN3 ADC reading = 374 (356 mV)
PIN3 ADC reading = 345 (328 mV)
PIN3 ADC reading = 472 (449 mV)
PIN3 ADC reading = 209 (199 mV)
PIN3 ADC reading = 567 (540 mV)
PIN3 ADC reading = 198 (188 mV)
PIN3 ADC reading = 585 (557 mV)
PIN3 ADC reading = 139 (132 mV)
PIN3 ADC reading = 644 (613 mV)
PIN3 ADC reading = 108 (102 mV)
PIN3 ADC reading = 635 (605 mV)
PIN3 ADC reading = 201 (191 mV)
PIN3 ADC reading = 555 (528 mV)
PIN3 ADC reading = 269 (256 mV)
PIN3 ADC reading = 485 (462 mV)
PIN3 ADC reading = 366 (348 mV)
PIN3 ADC reading = 342 (325 mV)
PIN3 ADC reading = 473 (450 mV)
PIN3 ADC reading = 212 (202 mV)
PIN3 ADC reading = 563 (536 mV)
PIN3 ADC reading = 211 (201 mV)
PIN3 ADC reading = 502 (478 mV)
PIN3 ADC reading = 173 (164 mV)
PIN3 ADC reading = 567 (540 mV)
PIN3 ADC reading = 101 (96 mV)
PIN3 ADC reading = 531 (505 mV)
PIN3 ADC reading = 201 (191 mV)
```

I'm measuring a value of a battery between 1.5 and 2.1V (with the voltage divider) and these values seem not accurate enough for Attenuation 11 db.

**Comment by katyo** [comment]
@elpiel Hmm, sounds strange.

Are you sure the configuration is correct? Is the correct GPIO is used? Have you try test with ESP IDF?
Looks like hardware issues.
I tested this PR with different `esp32c3`-based boards using different GPIOs, analog sources and attenuation configs. This works fine for me.

**Comment by elpiel** [comment]
I've tested this adc_cal on 2 different board (Espressif Rust dev board & Olimex dev board) yesterday and I got the same results.

When connecting the pin to ground I do get a very low values between 4-6.
Most readings are below a 1000 on the pin.

> Have you try test with ESP IDF?
No, I have not.

> Which features you use embassy-time-systick or embassy-time-timg0?
My Cargo.toml has:
```
default = ["embassy-time-timg0"]

embassy-time-timg0 = [
    "hal/embassy-time-timg0",
    "embassy-time/tick-hz-1_000_000",
]
```

**Comment by katyo** [comment]
@elpiel Possibly calibration is not a root of problem in your case.

Could you test without calibration? (`adc` example or `adc_cal` with `()` for `AdcCal` to bypass calibration)

**Comment by bugadani** [comment]
I see similarly unstable results on an ESP32-S3. ADC values are all over the place, averaging roughly about the correct value. I'm not sure if this is an issue with the 100kOhm+ source impedance combined with the undocumented input impedance of the ADC, or if it's noise coupled in from USB, a combination of the two or some third problem.

**Comment by katyo** [comment]
@elpiel The fact is that ADC input impedance on Espressif SoCs relatively low compared to many other MCUs (KOhms not MOhms).
I recommend you try use 10x-100x lower resistor values and add 10-100nF capacitor to input (after divider). You may need experimenting with it to get stable results.

**Comment by elpiel** [comment]
Testing on the Olimex board with soldered jumpers for the battery meassurement on GPIO 3 indeed shows good results!
I need to double check the values with my multimeter and figure out a way to not charge the battery via USB when measuring it.

https://raw.githubusercontent.com/OLIMEX/ESP32-C3-DevKit-Lipo/main/HARDWARE/ESP32-C3-DevKit-Lipo_Rev_A/ESP32-C3-DevKit-Lipo_Rev_A.pdf

On the embassy project I still get a panic but I will check whether this is because of unhappy embassy crate versions

File: esp-hal-common/src/analog/adc/cal_curve.rs
Comment (bugadani) [inline esp-hal-common/src/analog/adc/cal_curve.rs:48]: Is this comment a leftover from somewhere? I don't know what the magic number 2 is, or what `adc_error_coef_atten` refers to.

File: esp-hal-common/src/analog/adc/cal_line.rs
Comment (bugadani) [inline esp-hal-common/src/analog/adc/cal_line.rs:13]:
```suggestion
/// Marker trait for ADC units which support line fitting
```

File: esp-hal-common/src/analog/adc/cal_line.rs
Comment (bugadani) [inline esp-hal-common/src/analog/adc/cal_line.rs:15]:
```suggestion
/// Usually it means that reference points are stored in efuse.
```

File: esp-hal-common/src/soc/esp32c2/efuse.rs
Comment (bugadani) [inline esp-hal-common/src/soc/esp32c2/efuse.rs:58]: This `1` needs to be a named constant IMO.

File: esp-hal-common/src/soc/esp32c2/efuse.rs
Comment (bugadani) [inline esp-hal-common/src/soc/esp32c2/efuse.rs:70]: This `1` needs to be the same constant as above, I think.

File: esp-hal-common/src/soc/esp32c3/efuse.rs
Comment (bugadani) [inline esp-hal-common/src/soc/esp32c3/efuse.rs:59]: Same here

File: esp-hal-common/src/soc/esp32c3/efuse.rs
Comment (bugadani) [inline esp-hal-common/src/soc/esp32c3/efuse.rs:71]: Same here

File: esp-hal-common/src/soc/esp32c3/efuse.rs
Comment (bugadani) [inline esp-hal-common/src/soc/esp32c3/efuse.rs:104]: Same here

File: esp-hal-common/src/analog/adc/cal_line.rs
Comment (bugadani) [inline esp-hal-common/src/analog/adc/cal_line.rs:30]:
```suggestion
/// A reference point is a pair of a reference voltage and the corresponding mean
/// raw digital ADC value. Such values are usually stored in efuse bit fields
/// for each supported attenuation.
```

File: esp-hal-common/src/analog/adc/cal_line.rs
Comment (bugadani) [inline esp-hal-common/src/analog/adc/cal_line.rs:21]: This comment also needs to be updated, and probably turned into a proper doc comment?

File: esp-hal-common/src/analog/adc/cal_curve.rs
Comment (bugadani) [inline esp-hal-common/src/analog/adc/cal_curve.rs:56]: I don't think this shortcut is very useful.

File: esp-hal-common/src/analog/adc/cal_line.rs
Comment (bugadani) [inline esp-hal-common/src/analog/adc/cal_line.rs:48]: Just a code style nit: this would look better as:
```suggestion
impl<ADCI> AdcCalScheme<ADCI> for AdcCalLine<ADCI>
where
    ADCI: AdcCalEfuse + AdcHasLineCal + RegisterAccess
{
```

**Review by bugadani — COMMENTED** [review]
I would like to see the eFuse calibration versions named (a plain `1` is a bit confusing to read), but apart from that, this PR is in a really promising state.

I can see a few grammar issues (mine is probably incorrect, too!), and a few comments that look like they were copied from C or written to explain the C code during development. Really nothing major.

It would be nice if this PR could be merged in the near future 🙏

**Review by MabezDev — CHANGES_REQUESTED** [review]

File: esp-hal-common/src/analog/adc/cal_curve.rs
Comment (katyo) [reply esp-hal-common/src/analog/adc/cal_curve.rs:48]: Oh, seems I copied this comment from C source.

File: esp-hal-common/src/soc/esp32c2/efuse.rs
Comment (katyo) [reply esp-hal-common/src/soc/esp32c2/efuse.rs:58]: I would like avoid introduce new constants.
Ok, I'll think about it.

File: esp-hal-common/src/analog/adc/cal_curve.rs
Comment (bugadani) [reply esp-hal-common/src/analog/adc/cal_curve.rs]: Hmm I might not have been clear with this comment. What I means was the `== 0 { // early return` part, as it's probably not a very common case that the ADC measures an exact 0. My apologies.

It's a bit hard to see what happened to this function, because you force pushed. Did you inline `calc_error` into `adc_val`, or do I just misremember how this was organized before?

**Comment by bugadani** [comment]
I just want to say that I'm writing on implementing this PR for the S3, just so others don't waste their time on that :) I'll post the results once this one gets merged.

**Review by MabezDev — APPROVED** [review]
After thinking some more, I don't think the consts are that big of a deal, we can always reference the ESP-IDF code to actually understand what's going on.

I think a quick rebase on main and this should be good to go from my perspective :).

**Comment by bugadani** [comment]
> After thinking some more, I don't think the consts are that big of a deal, we can always reference the ESP-IDF code to actually understand what's going on.

Honestly, I think the `if major == 0 { 1 }` pattern deserves some sort of explanation (be that a comment or a named constant) in the code as it looks like a mistake.

**Review by jessebraham — APPROVED** [review]
I think we could introduce some new constants here, as others have suggested, but in the interest of getting this PR merged I'm gonna say LGTM and we can come back to it at a later date. I also noticed that the H2 is missing from this, but again that can wait. I will create an issue to remind us, and may end up just tackling that myself.

Thanks for all your hard work on this, will be great to actually be able to use our ADCs now! 😁

**Comment by katyo** [comment]
@jessebraham Because esp-idf still does not support calibration for H2 so I don't know how we can add support for it just now.
(see [efuse/esp32h2/esp_efuse_rtc_calib.c#L13](https://github.com/espressif/esp-idf/blob/903af13e847cd301e476d8b16b4ee1c21b30b5c6/components/efuse/esp32h2/esp_efuse_rtc_calib.c#L13))

**Pull request merged**
